### PR TITLE
feat(opensearch): add support for latest amazon opensearch service 2.3

### PR DESCRIPTION
**Pull request by agusti-t**
fixes https://github.com/aws/aws-cdk/issues/22942 
Description:
AWS Launched the support for Amazon OpenSearch Service 2.3 as per the announcement here: https://aws.amazon.com/about-aws/whats-new/2022/11/amazon-opensearch-service-supports-opensearch-version-2-3/

AWS CDK does not support the latest version yet, and can not create an OpenSearch cluster with CDK using 2.3.

----

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md)

### Adding new Unconventional Dependencies:

* [ ] This PR adds new unconventional dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md/#adding-new-unconventional-dependencies)

### New Features

* [ ] Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/main/INTEGRATION_TESTS.md)?
	* [ ] Did you use `yarn integ` to deploy the infrastructure and generate the snapshot (i.e. `yarn integ` without `--dry-run`)?

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Review by TheRealAmazonKendra — DISMISSED** [review]
Thank you for your contribution! 

The minimal changes in the snapshot of the integration test usually indicates that the tests weren't actually run. Please revert the change  in the snapshot and run `yarn integ --update-on-failed` to validate this change and to produce the new snapshots. I also notice that the integ test isn't using the new [IntegTest](https://github.com/aws/aws-cdk/blob/main/INTEGRATION_TESTS.md) construct. Please update that.

**Comment by agusti-t** [comment]
> Thank you for your contribution!
> 
> The minimal changes in the snapshot of the integration test usually indicates that the tests weren't actually run. Please revert the change in the snapshot and run `yarn integ --update-on-failed` to validate this change and to produce the new snapshots. I also notice that the integ test isn't using the new [IntegTest](https://github.com/aws/aws-cdk/blob/main/INTEGRATION_TESTS.md) construct. Please update that.

Thanks for the review, I tried using the new `IntegTest` and run the tests as you said and run into this problem:

```
➜  aws-cdk git:(OpenSearch_2_3) ✗ yarn integ-runner --directory packages/@aws-cdk/aws-appsync --update-on-failed
yarn run v1.22.19
$ /Users/agusti/src/aws-cdk/node_modules/.bin/integ-runner --directory packages/@aws-cdk/aws-appsync --update-on-failed

Verifying integration test snapshots...

  UNCHANGED  integ.appsync-none 0.915s
  UNCHANGED  integ.log-retention 1.066s
  UNCHANGED  integ.api-import 1.092s
  UNCHANGED  integ.appsync-lambda 1.082s
  CHANGED    integ.graphql-opensearch 1.214s
      appsync-opensearch exists in snapshot, but not in actual
  CHANGED    integ.graphql-opensearch 1.214s
      OpensSearch23Stack does not exist in snapshot, but does in actual
  UNCHANGED  integ.auth-apikey 1.92s
  UNCHANGED  integ.graphql-schema 1.919s
  UNCHANGED  integ.graphql-iam 1.948s
  UNCHANGED  integ.graphql-elasticsearch 1.972s
  UNCHANGED  integ.graphql 2.045s

Snapshot Results:

Tests:    1 failed, 10 total

Running integration tests for failed tests...

Running in parallel across regions: us-east-1, us-east-2, us-west-2
Running test /Users/agusti/src/aws-cdk/packages/@aws-cdk/aws-appsync/test/integ.graphql-opensearch.js in us-east-1
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
!!                                                                                  !!
!!  This software has not been tested with node v19.0.1.                            !!
!!  You may to encounter runtime issues, and should switch to a supported release.  !!
!!                                                                                  !!
!!  This software is currently running on node v19.0.1.                             !!
!!  As of the current release of this software, supported node releases are:        !!
!!  - ^18.0.0 (Planned end-of-life: 2025-04-30)                                     !!
!!  - ^16.3.0 (Planned end-of-life: 2023-09-11)                                     !!
!!  - ^14.6.0 (Planned end-of-life: 2023-04-30)                                     !!
!!                                                                                  !!
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

✨  Synthesis time: 0.3s

OpensSearch23Stack: building assets...

[0%] start: Building 7725c14cf18390409c0bad4e2d2037a44b300aac680ce62082d410f076e7f331:current_account-current_region
[100%] success: Built 7725c14cf18390409c0bad4e2d2037a44b300aac680ce62082d410f076e7f331:current_account-current_region

OpensSearch23Stack: assets built

lambda-api-latebound-deploymentstage/DefaultTest/DeployAssert: building assets...

[0%] start: Building 21fbb51d7b23f6a6c262b46a9caee79d744a3ac019fd45422d988b96d44b2a22:current_account-current_region
[100%] success: Built 21fbb51d7b23f6a6c262b46a9caee79d744a3ac019fd45422d988b96d44b2a22:current_account-current_region

lambda-api-latebound-deploymentstage/DefaultTest/DeployAssert: assets built

OpensSearch23Stack
OpensSearch23Stack: deploying...
[0%] start: Publishing 7725c14cf18390409c0bad4e2d2037a44b300aac680ce62082d410f076e7f331:current_account-current_region
[100%] success: Published 7725c14cf18390409c0bad4e2d2037a44b300aac680ce62082d410f076e7f331:current_account-current_region
OpensSearch23Stack: creating CloudFormation changeset...
OpensSearch23Stack | 0/10 | 12:38:57 AM | REVIEW_IN_PROGRESS   | AWS::CloudFormation::Stack     | OpensSearch23Stack User Initiated
OpensSearch23Stack | 0/10 | 12:39:04 AM | CREATE_IN_PROGRESS   | AWS::CloudFormation::Stack     | OpensSearch23Stack User Initiated
OpensSearch23Stack | 0/10 | 12:39:09 AM | CREATE_IN_PROGRESS   | AWS::AppSync::GraphQLApi       | api (apiC8550315)
OpensSearch23Stack | 0/10 | 12:39:09 AM | CREATE_IN_PROGRESS   | AWS::IAM::User                 | User (User00B015A1)
OpensSearch23Stack | 0/10 | 12:39:10 AM | CREATE_IN_PROGRESS   | AWS::IAM::Role                 | api/ds/ServiceRole (apidsServiceRoleBDB08107)
OpensSearch23Stack | 0/10 | 12:39:10 AM | CREATE_IN_PROGRESS   | AWS::IAM::User                 | User (User00B015A1) Resource creation Initiated
OpensSearch23Stack | 0/10 | 12:39:10 AM | CREATE_FAILED        | AWS::IAM::Role                 | api/ds/ServiceRole (apidsServiceRoleBDB08107) Invalid principal in policy: "SERVICE":"appsync" (Service: AmazonIdentityManagement; Status Code: 400; Error Code: MalformedPolicyDocument; Request ID: d3941825-78dd-4440-9132-0254299af33f; Proxy: null)
OpensSearch23Stack | 0/10 | 12:39:13 AM | CREATE_IN_PROGRESS   | AWS::AppSync::GraphQLApi       | api (apiC8550315) Resource creation Initiated
OpensSearch23Stack | 1/10 | 12:39:13 AM | CREATE_COMPLETE      | AWS::AppSync::GraphQLApi       | api (apiC8550315)
OpensSearch23Stack | 1/10 | 12:39:16 AM | CREATE_IN_PROGRESS   | AWS::AppSync::GraphQLSchema    | api/Schema (apiSchema0EA92056)
OpensSearch23Stack | 1/10 | 12:39:18 AM | CREATE_IN_PROGRESS   | AWS::AppSync::GraphQLSchema    | api/Schema (apiSchema0EA92056) Resource creation Initiated
OpensSearch23Stack | 2/10 | 12:39:46 AM | CREATE_COMPLETE      | AWS::IAM::User                 | User (User00B015A1)
OpensSearch23Stack | 2/10 | 12:39:47 AM | CREATE_IN_PROGRESS   | AWS::OpenSearchService::Domain | Domain (Domain66AC69E0)
OpensSearch23Stack | 2/10 | 12:39:50 AM | CREATE_IN_PROGRESS   | AWS::OpenSearchService::Domain | Domain (Domain66AC69E0) Resource creation Initiated
OpensSearch23Stack | 3/10 | 12:40:19 AM | CREATE_COMPLETE      | AWS::AppSync::GraphQLSchema    | api/Schema (apiSchema0EA92056)
OpensSearch23Stack | 3/10 | 12:40:22 AM | CREATE_IN_PROGRESS   | AWS::AppSync::ApiKey           | api/DefaultApiKey (apiDefaultApiKey6AB8D7C4)
OpensSearch23Stack | 3/10 | 12:40:24 AM | CREATE_IN_PROGRESS   | AWS::AppSync::ApiKey           | api/DefaultApiKey (apiDefaultApiKey6AB8D7C4) Resource creation Initiated
OpensSearch23Stack | 4/10 | 12:40:24 AM | CREATE_COMPLETE      | AWS::AppSync::ApiKey           | api/DefaultApiKey (apiDefaultApiKey6AB8D7C4)
OpensSearch23Stack | 5/10 | 12:56:02 AM | CREATE_COMPLETE      | AWS::OpenSearchService::Domain | Domain (Domain66AC69E0)
OpensSearch23Stack | 5/10 | 12:56:03 AM | CREATE_FAILED        | AWS::CloudFormation::Stack     | OpensSearch23Stack The following resource(s) failed to create: [apidsServiceRoleBDB08107].

Failed resources:
OpensSearch23Stack | 12:39:10 AM | CREATE_FAILED        | AWS::IAM::Role                 | api/ds/ServiceRole (apidsServiceRoleBDB08107) Invalid principal in policy: "SERVICE":"appsync" (Service: AmazonIdentityManagement; Status Code: 400; Error Code: MalformedPolicyDocument; Request ID: d3941825-78dd-4440-9132-0254299af33f; Proxy: null)

 ❌  OpensSearch23Stack failed: Error: The stack named OpensSearch23Stack failed to deploy: CREATE_FAILED (The following resource(s) failed to create: [apidsServiceRoleBDB08107]. ): Invalid principal in policy: "SERVICE":"appsync" (Service: AmazonIdentityManagement; Status Code: 400; Error Code: MalformedPolicyDocument; Request ID: d3941825-78dd-4440-9132-0254299af33f; Proxy: null)
    at FullCloudFormationDeployment.monitorDeployment (/Users/agusti/src/aws-cdk/packages/aws-cdk/lib/api/deploy-stack.ts:505:13)
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
    at deployStack (/Users/agusti/src/aws-cdk/packages/aws-cdk/lib/cdk-toolkit.ts:264:24)
    at /Users/agusti/src/aws-cdk/packages/aws-cdk/lib/deploy.ts:39:11
    at run (/Users/agusti/src/aws-cdk/node_modules/p-queue/dist/index.js:163:29)

 ❌ Deployment failed: Error: Stack Deployments Failed: Error: The stack named OpensSearch23Stack failed to deploy: CREATE_FAILED (The following resource(s) failed to create: [apidsServiceRoleBDB08107]. ): Invalid principal in policy: "SERVICE":"appsync" (Service: AmazonIdentityManagement; Status Code: 400; Error Code: MalformedPolicyDocument; Request ID: d3941825-78dd-4440-9132-0254299af33f; Proxy: null)
    at Object.exports.deployStacks (/Users/agusti/src/aws-cdk/packages/aws-cdk/lib/deploy.ts:61:11)
    at CdkToolkit.deploy (/Users/agusti/src/aws-cdk/packages/aws-cdk/lib/cdk-toolkit.ts:338:7)
    at initCommandLine (/Users/agusti/src/aws-cdk/packages/aws-cdk/lib/cli.ts:364:12)

Stack Deployments Failed: Error: The stack named OpensSearch23Stack failed to deploy: CREATE_FAILED (The following resource(s) failed to create: [apidsServiceRoleBDB08107]. ): Invalid principal in policy: "SERVICE":"appsync" (Service: AmazonIdentityManagement; Status Code: 400; Error Code: MalformedPolicyDocument; Request ID: d3941825-78dd-4440-9132-0254299af33f; Proxy: null)
  FAILED     integ.graphql-opensearch-lambda-api-latebound-deploymentstage/DefaultTest (undefined/us-east-1) 2389.226s
      Integration test failed: Error: Command exited with status 1

Test Results:

Tests:    1 failed, 1 total
Error: Some integration tests failed!
    at main (/Users/agusti/src/aws-cdk/packages/@aws-cdk/integ-runner/lib/cli.js:114:23)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

Any idea what it could be? I tried looking it up without being able to find anything.

**Comment by TheRealAmazonKendra** [comment]
@Mergifyio update

**Review by TheRealAmazonKendra — DISMISSED** [review]
Please see my comments inline.

We had a bug that was causing your test failures. I updated your branch to be up-to-date with `main` so the fix is in there now. You should now be able to run `yarn integ --update-on-failed` and get the correct output. Note that you may need to clean the generated files from the build because the test runner is expecting to see certain files that don't exist due to the update in stack name and using the new `IntegTest` construct. 

Please let me know if you have any questions about this.

**Comment by agusti-t** [comment]
@TheRealAmazonKendra Trying again, it took me a while to be able to run the tests with `yarn integ --update-on-failed`, it kept complaining about not finding the `integ` cmd, TBH I don't know how/why it started working. I had to keep trying with `yarn integ-runner --update-on-failed` and that always failed. It could also be that I rebased `main` and got the bug (suspect [this](https://github.com/aws/aws-cdk/commit/0bfce8965ee50ab79054e6f5a4c6bbecb0955e19) one) fix in, don't understand why I didn't get it earlier.

Anyway, it worked on my machine now and I committed everything, but I ask you to please check commits ([1](https://github.com/aws/aws-cdk/pull/22943/commits/544911ce7726f4a640382bbd0024ac09b9082cae), [2](https://github.com/aws/aws-cdk/pull/22943/commits/56ec81f4878d86f234cca0f122e4824faa2fde2a)) because I am not sure if these files should be committed. Of course, the rest of the commits should be checked too.

**Comment by Naumel** [comment]
@Mergifyio update

**Comment by agusti-t** [comment]
@Naumel thanks for merging, LMK if I can do anything from my side. I do have some unanswered outstanding questions in [this](https://github.com/aws/aws-cdk/pull/22943#issuecomment-1326791311) comment. Can you help?

**Comment by Naumel** [comment]
https://github.com/aws/aws-cdk/pull/22943#issuecomment-1342855326

Hello,

I tried to unblock the merge from `main` and I think I made it more complicated instead of helpful.
I'll get the integ test back in shape, once the build is up and running, **all** content will be checked.

**Review by rix0rrr — DISMISSED** [review]

**Comment by agusti-t** [comment]
@rix0rrr should I try to merge or will you handle that? What about [my questions](https://github.com/aws/aws-cdk/pull/22943#issuecomment-1326791311)? Are they relevant?

**Comment by agusti-t** [comment]
@rix0rrr and @Naumel FYI, I just rebased latest from master and force pushed, had a conflict. I didn't re run the integration test however.

**Comment by Naumel** [comment]
This PR has too many cooks, I'm bowing out.

```
@aws-cdk/aws-appsync: Snapshot Results: 
@aws-cdk/aws-appsync: Tests:    5 failed, 9 total
@aws-cdk/aws-appsync: Error: Some tests failed!
```

**Comment by TheRealAmazonKendra** [comment]
@Mergifyio update

**Comment by TheRealAmazonKendra** [comment]
> @TheRealAmazonKendra Trying again, it took me a while to be able to run the tests with `yarn integ --update-on-failed`, it kept complaining about not finding the `integ` cmd, TBH I don't know how/why it started working. I had to keep trying with `yarn integ-runner --update-on-failed` and that always failed. It could also be that I rebased `main` and got the bug (suspect [this](https://github.com/aws/aws-cdk/commit/0bfce8965ee50ab79054e6f5a4c6bbecb0955e19) one) fix in, don't understand why I didn't get it earlier.
> 
> Anyway, it worked on my machine now and I committed everything, but I ask you to please check commits ([1](https://github.com/aws/aws-cdk/pull/22943/commits/544911ce7726f4a640382bbd0024ac09b9082cae), [2](https://github.com/aws/aws-cdk/pull/22943/commits/56ec81f4878d86f234cca0f122e4824faa2fde2a)) because I am not sure if these files should be committed. Of course the rest of the commits should be checked too.

On your first question, were you potentially trying to run the integ command from the package root instead of from inside the opensearch module? That's the only thing that I can think of that would cause that. Otherwise, I'm kind of stumped.

On why you didn't get the error earlier, my best guess is that when you cloned the package, the error wasn't in it yet and then upon rebasing from main, the buggy code had been merged at that point.

All speculation on my part but that's about all I can do ¯\_(ツ)_/¯ .

If the tests are still failing at this point, I can help correct them. Let's see what this build does.

**Comment by dblock** [comment]
@agusti-t can you please close this PR so we can get https://github.com/aws/aws-cdk/pull/23621#issuecomment-1379684041 merged? Thanks

**Comment by dblock** [comment]
@TheRealAmazonKendra with Mergifyio you just updated a PR that was abandoned for 3 weeks - will we have to wait another 3 week for the bot to close it?

**Comment by agusti-t** [comment]
> @agusti-t can you please close this PR so we can get [#23621 (comment)](https://github.com/aws/aws-cdk/pull/23621#issuecomment-1379684041) merged? Thanks

Well technically it wasn't abandoned. I was just waiting for answers to my questions in https://github.com/aws/aws-cdk/pull/22943#issuecomment-1326791311.

**Comment by agusti-t** [comment]
> If the tests are still failing at this point, I can help correct them. Let's see what this build does.

@TheRealAmazonKendra fails, could you help please?

**Comment by dblock** [comment]
@agusti-t changes here are way bigger than just adding support for OpenSearch 2.3. Can we maybe please split the changes? Feel free to do it yourself - the minimum is https://github.com/aws/aws-cdk/pull/23621 which just adds the version, and it passes tests. Then you can try making all the other updates you have here on top? It’s all yours now.

**Comment by agusti-t** [comment]
> @agusti-t changes here are way bigger than just adding support for OpenSearch 2.3. The app sync changes are not needed for that. Can we maybe please split the changes? Feel free to do it yourself - the minimum is #23621 which just adds the version, and it passes tests. Then you can try making all the other updates you have here on top? It’s all yours now - I would close this PR, open one that looks just like 23621 and then try to make app sync updates separately.

I'm all for that, having support for 2.3 I mean. Let's see if I get help with my questions in a couple of days, otherwise I'll close and this and copy your PR (or I can let you know and you can open it yourself).

**Comment by dblock** [comment]
I believe tests are failing because you're making changes that would be destructive during an upgrade. I'm not sure what these other changes in appsync do, but they are not related to OpenSearch 2.3 update which was the purpose of this PR.

```
@aws-cdk/aws-appsync:     Stack: aws-appsync-integ - Resource: ApiorderDsQuerygetCustomerOrdersEqResolverEF9D5350 - Impact: WILL_DESTROY
@aws-cdk/aws-appsync:     Stack: aws-appsync-integ - Resource: ApipaymentDsQuerygetPaymentResolverD172BFC9 - Impact: WILL_DESTROY
```

**Comment by TheRealAmazonKendra** [comment]
> @TheRealAmazonKendra with Mergifyio you just updated a PR that was abandoned for 3 weeks - will we have to wait another 3 week for the bot to close it?

Four weeks, technically, but this PR was not abandoned. It was awaiting answers from our team. The appsync tests are irrelevant to this PR altogether. They got pulled in here as a result of a bug that we had in our codebase when the PR was opened that was causing the integ tests to fail. I will clean up this part. 

Once I've done so, I'll update with a new review.

**Comment by TheRealAmazonKendra** [comment]
I think the only thing that's going to be missing from this PR is unit tests.

**Review by TheRealAmazonKendra — DISMISSED** [review]
Assuming that this build passes, I just need you to add some unit tests to cover the new version and I can approve this.

**Comment by agusti-t** [comment]
> Contributor

I just bumped the `defaultVersion` to 2.3 and adapted existing tests. Don't know if that is enough, but if not, LMK what you'd like to see in terms of unit tests and I'll do my best. I'm a noob when it comes to ts.

**Comment by dblock** [comment]
Thank you @agusti-t !

**Review by TheRealAmazonKendra — DISMISSED** [review]

File: packages/@aws-cdk/aws-opensearchservice/test/domain.test.ts [hunk 1]
```diff
@@ -29,7 +29,7 @@ const readWriteActions = [
   ...writeActions,
 ];
 
-const defaultVersion = EngineVersion.OPENSEARCH_1_0;
+const defaultVersion = EngineVersion.OPENSEARCH_2_3;
```
[the remark this answers is not in the document]
Comment (agusti-t) [reply packages/@aws-cdk/aws-opensearchservice/test/domain.test.ts]: Working on it, will request review when ready.

[the remark this answers is not in the document]
Comment (agusti-t) [reply packages/@aws-cdk/aws-opensearchservice/test/domain.test.ts]: Done, I think it's better now. LMK what you think.

File: packages/@aws-cdk/aws-opensearchservice/test/domain.test.ts [hunk 2]
```diff
@@ -183,8 +190,14 @@ test('grants kms permissions if needed', () => {
 
 });
 
-test('minimal example renders correctly', () => {
-  new Domain(stack, 'Domain', { version: defaultVersion });
+each([
```
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/aws-opensearchservice/test/domain.test.ts:193]: Fantastic, thank you!

**Review by TheRealAmazonKendra — APPROVED** [review]

**Pull request merged**